crates/red_knot_python_semantic/src/semantic_index/use_def/bitset.rs

```diff
@@ -0,0 +1,402 @@
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(super) enum BitSetArray<const B: usize, const N: usize> {
+    Inline {
+        /// Array of N [`BitSet`].
+        array: [BitSet<B>; N],
+
+        /// How many of the bitsets are used?
+        size: usize,
+    },
+
+    Heap(Vec<BitSet<B>>),
+}
+
+impl<const B: usize, const N: usize> Default for BitSetArray<B, N> {
+    fn default() -> Self {
+        Self::Inline {
+            array: std::array::from_fn(|_| BitSet::default()),
+            size: 0,
+        }
+    }
+}
+
+impl<const B: usize, const N: usize> BitSetArray<B, N> {
+    /// Create a [`BitSetArray`] of `size` empty [`BitSet`]s.
+    pub(super) fn of_size(size: usize) -> Self {
+        let mut array = Self::default();
+        for _ in 0..size {
+            array.push(BitSet::default());
+        }
+        array
+    }
+
+    fn overflow(&mut self) {
+        match self {
+            Self::Inline { array, size } => {
+                let mut vec: Vec<BitSet<B>> = vec![];
+                for bitset in array.iter().take(*size) {
+                    vec.push(bitset.clone());
+                }
+                *self = Self::Heap(vec);
+            }
+            Self::Heap(_) => {}
+        }
+    }
+
+    /// Push a [`BitSet`] onto the end of the array.
+    pub(super) fn push(&mut self, new: BitSet<B>) {
+        match self {
+            Self::Inline { array, size } => {
+                *size += 1;
+                if *size > N {
+                    self.overflow();
+                    self.push(new);
+                } else {
+                    array[*size - 1] = new;
+                }
+            }
+            Self::Heap(vec) => vec.push(new),
+        }
+    }
+
+    /// Return a mutable reference to the last [`BitSet`] in the array, or None.
+    pub(super) fn last_mut(&mut self) -> Option<&mut BitSet<B>> {
+        match self {
+            Self::Inline { array, size } => {
+                if *size == 0 {
+                    None
+                } else {
+                    Some(&mut array[*size - 1])
+                }
+            }
+            Self::Heap(vec) => vec.last_mut(),
+        }
+    }
+
+    /// Insert `value` into every [`BitSet`] in this [`BitSetArray`].
+    pub(super) fn insert_in_each(&mut self, value: u32) {
+        match self {
+            Self::Inline { array, size } => {
+                for bitset in array.iter_mut().take(*size) {
+                    bitset.insert(value);
+                }
+            }
+            Self::Heap(vec) => {
+                for bitset in vec {
+                    bitset.insert(value);
+                }
+            }
+        }
+    }
+
+    /// Return an iterator over each [`BitSet`] in this [`BitSetArray`].
+    pub(super) fn iter(&self) -> BitSetArrayIterator<'_, B, N> {
+        match self {
+            Self::Inline { array, size } => BitSetArrayIterator::Inline {
+                array,
+                index: 0,
+                size: *size,
```

Comment:
Could we slice the array instead and directly return an `std::slice::Iter`?
```suggestion
                &array[..*size],
```

Reply:
Good idea, will try that!

Reply:
Great suggestion, this allowed totally removing `BitSetArrayIterator`.